### PR TITLE
[Feature] add scannet instance dataset with metrics

### LINKED ISSUE #1204: Instance segmentation on ScanNet
HI,

I've adapted `ScanNetSegDataset` to `ScanNetInstanceSegDataset` with (almost) original ScanNet evaluator. Will it be interesting for the community? If yes, to which branch (`master` or `v1-dev`) should I make a PR?

### PR DESCRIPTION
**Pull request by filaPro**
Add ScanNet instance segmentation dataset with evaluator. Close #1204.

Can you please have a look in general structure (e.g. do we need separate `INSTANCE_SEG_DATASETS` in registry, can `ScanNetInstanceSegDataset` be inherited from `Custom3DSegDataset`, do we load only `with_mask_3d` and `with_seg_3d` in `LoadAnnotations3D`, etc...)

Also lint is broken probably not by this PR.

### CODE REVIEW

**Comment by ZCMax** [comment]
Yes, the lint broken problem will be fixed in this PR https://github.com/open-mmlab/mmdetection3d/pull/1166/files

File: mmdet3d/datasets/scannet_dataset.py [hunk 1]
```diff
@@ -460,3 +460,149 @@ def format_results(self, results, txtfile_prefix=None):
...
+               'garbagebin')
+
+    VALID_CLASS_IDS = (3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 16, 24, 28, 33, 34,
+                       36, 39)
+
+    ALL_CLASS_IDS = tuple(range(41))
+
+    def get_ann_info(self, index):
+        """Get annotation info according to the given index.
+        Args:
```
Comment (Tai-Wang) [inline mmdet3d/datasets/scannet_dataset.py:480]: Need a blank line before Args and Returns

Comment (Tai-Wang) [reply mmdet3d/datasets/scannet_dataset.py:480]: Please also check other places

File: mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# adapted from https://github.com/ScanNet/ScanNet/blob/master/BenchmarkScripts/3d_evaluation/evaluate_semantic_instance.py # noqa
+import numpy as np
+from copy import deepcopy
+
+from . import util_3d
+
+
+def evaluate_matches(matches, class_labels, options):
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py:9]: Please add docstring to important functions (and also check other functions)

File: mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    # results: class x overlap
+    ap = np.zeros((len(dist_threshes), len(class_labels), len(overlaps)),
+                  np.float)
+    for di, (min_region_size, distance_thresh, distance_conf) in enumerate(
+            zip(min_region_sizes, dist_threshes, dist_confs)):
+        for oi, overlap_th in enumerate(overlaps):
+            pred_visited = {}
+            for m in matches:
+                # (@filaPro:) it is not used
+                # for p in matches[m]['pred']:
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py:24]: Please clean unused codes.
What do you mean by commenting `@filaPro`? Do they also need cleaning?

File: mmdet3d/core/evaluation/instance_seg_eval.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    Args:
+        gt_semantic_masks (list[torch.Tensor]): Ground truth semantic masks.
+        gt_instance_masks (list[torch.Tensor]): Ground truth instance masks.
+        pred_instance_masks (list[torch.Tensor]): Predicted instance masks.
+        pred_instance_labels (list[torch.Tensor]): Predicted instance labels.
+        pred_instance_scores (list[torch.Tensor]): Predicted instance labels.
+        valid_class_ids (tuple[int]): Ids of valid categories.
+        class_labels (tuple[str]): Names of valid categories.
+        options (dict): Additional options. Keys may contain: `overlaps`,
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/instance_seg_eval.py:69]: dict, optional

Comment (Tai-Wang) [reply mmdet3d/core/evaluation/instance_seg_eval.py:69]: Default: None.

File: mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                    ap_current = float('nan')
+                ap[di, li, oi] = ap_current
+    return ap
+
+
+def compute_averages(aps, options, class_labels):
+    d_inf = 0
+    o50 = np.where(np.isclose(options['overlaps'], 0.5))
+    o25 = np.where(np.isclose(options['overlaps'], 0.25))
+    oAllBut25 = np.where(np.logical_not(np.isclose(options['overlaps'], 0.25)))
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py:185]: o_all_but25 would be more consistent with other variable names?

File: mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    avg_dict = {}
+    # avg_dict['all_ap'] = np.nanmean(aps[ d_inf, :, :])
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py:187]: Can be removed

File: mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    avg_dict['all_ap'] = np.nanmean(aps[d_inf, :, oAllBut25])
+    avg_dict['all_ap_50%'] = np.nanmean(aps[d_inf, :, o50])
+    avg_dict['all_ap_25%'] = np.nanmean(aps[d_inf, :, o25])
+    avg_dict['classes'] = {}
+    for (li, label_name) in enumerate(class_labels):
+        avg_dict['classes'][label_name] = {}
+        # avg_dict['classes'][label_name]['ap'] = np.average(aps[d_inf, li, :])
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py:194]: Can be removed together with the above one or not?

File: mmdet3d/core/evaluation/scannet_utils/util_3d.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# copied with deletions from https://github.com/ScanNet/ScanNet/blob/master/BenchmarkScripts/util_3d.py # noqa
+import json
+import numpy as np
+
+
+class Instance(object):
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/scannet_utils/util_3d.py:7]: Need docstring

**Review by Tai-Wang — COMMENTED** [review]
Awesome work! Overall looks good to me. Just need to fix the detailed comments.

File: mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+        avg_dict['classes'][label_name]['ap'] = np.average(aps[d_inf, li,
+                                                               o_all_but25])
+        avg_dict['classes'][label_name]['ap50%'] = np.average(aps[d_inf, li,
+                                                                  o50])
+        avg_dict['classes'][label_name]['ap25%'] = np.average(aps[d_inf, li,
+                                                                  o25])
+    return avg_dict
+
+
+def assign_instances_for_scan(pred_info, gt_ids, options, valid_class_ids,
```
Comment (ZCMax) [inline mmdet3d/core/evaluation/scannet_utils/evaluate_semantic_instance.py:220]: add docstrings here

**Comment by ZCMax** [comment]
Can you merge the our latest dev branch to resolve the link problem?

**Comment by filaPro** [comment]
Eventually I have broken this PR :( Should I close this one and open new?

**Comment by ZCMax** [comment]
> 

no, the new lint problem is caused by merging master branch to dev branch, we will fix it right now. After that, you can merge dev branch again.

**Comment by Tai-Wang** [comment]
Hi @filaPro sorry for the late reply. Could you please merge v1.0.0dev0? That should fix the lint and CI, and then we can merge this PR.

**Comment by filaPro** [comment]
Don't quite understand why build is failed now...

**Comment by filaPro** [comment]
Lint is fine. Pytest is also fine on my side. Maybe smth wrong with `mmcv` version on your side?

**Comment by Tai-Wang** [comment]
> Lint is fine. Pytest is also fine on my side. Maybe smth wrong with `mmcv` version on your side?

Yes, something related to spconv op is registered in mmcv now. Our op migration PR should fix this problem. Please don't worry.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**